File: ios/Sources/SentinelSDK/SentinelSDK.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

/// Primary facade for the Sentinel SDK on iOS.
public final class SentinelSDK {

    public static let shared = SentinelSDK()
    public static let version = "1.0.0"

    private var apiKey: String?
    private var environment: String = "production"
    private var isInitialized = false

    private let personalDataValidator = PersonalDataValidator()
    private let deviceDataCollector = DeviceDataCollector()
    private let locationCollector = LocationCollector()
    private let appDetectionCollector = AppDetectionCollector()

    private init() {}

    /// Initializes the Sentinel SDK with client API credentials.
    public func initialize(apiKey: String, environment: String = "production") {
        self.apiKey = apiKey
        self.environment = environment
        self.isInitialized = true
    }

    /// Configuration options for data capture.
    public struct CaptureConfiguration {
        public let userData: PersonalDataValidator.Input
        public let locationTimeoutSeconds: TimeInterval
        public let wrapper: String
        public let customTargets: [AppDetectionCollector.TargetApp]?

        public init(
            userData: PersonalDataValidator.Input,
            locationTimeoutSeconds: TimeInterval = 5.0,
            wrapper: String = "native",
            customTargets: [AppDetectionCollector.TargetApp]? = nil
        ) {
            self.userData = userData
            self.locationTimeoutSeconds = locationTimeoutSeconds
            self.wrapper = wrapper
            self.customTargets = customTargets
        }
    }

    /// Captures all required telemetry, runs validation, and generates a signed payload.
    public func capture(configuration: CaptureConfiguration) async -> SentinelPayload {
        let startTime = Date()

        // 1. Validate personal data
        let userData = personalDataValidator.validate(input: configuration.userData)

        // 2. Collect device telemetry
        let deviceData = await MainActor.run {
            self.deviceDataCollector.collect()
        }

        // 3. Collect high-accuracy location
        let locationData = await locationCollector.collectLocation(timeoutSeconds: configuration.locationTimeoutSeconds)

        // 4. Scan targeted applications on main actor
        let targets = configuration.customTargets ?? AppDetectionCollector.defaultTargets
        let installedApps = await MainActor.run {
            self.appDetectionCollector.scan(targets: targets)
        }

        let captureDurationMs = max(1, Int(Date().timeIntervalSince(startTime) * 1000))

        // 5. Build metadata
        #if os(iOS)
        let osVersion = await MainActor.run { UIDevice.current.systemVersion }
        #else
        let osVersion = "unknown"
        #endif

        let metadata = SentinelMetadata(
            sdkVersion: SentinelSDK.version,
            platform: "ios",
            osVersion: osVersion,
            wrapper: configuration.wrapper,
            timestamp: ISO8601DateFormatter().string(from: startTime),
            captureDurationMs: captureDurationMs
        )

        // 6. Assemble payload data for signing
        let unsignedPayload = PreSecurityPayload(
            metadata: metadata,
            userData: userData,
            deviceData: deviceData,
            locationData: locationData,
            installedApps: installedApps
        )

        let signer = PayloadSigner(secretKey: apiKey)
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        let payloadData = (try? encoder.encode(unsignedPayload)) ?? Data()
        let securityData = signer.sign(data: payloadData)

        return SentinelPayload(
            metadata: metadata,
            userData: userData,
            deviceData: deviceData,
            locationData: locationData,
            installedApps: installedApps,
            security: securityData
        )
    }

    /// Captures and serializes the payload directly to a formatted JSON string.
    public func captureJSON(configuration: CaptureConfiguration) async throws -> String {
        let payload = await capture(configuration: configuration)
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(payload)
        guard let jsonString = String(data: data, encoding: .utf8) else {
            throw NSError(domain: "SentinelSDK", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to encode JSON payload"])
        }
        return jsonString
    }
}

private struct PreSecurityPayload: Codable {
    let metadata: SentinelMetadata
    let userData: SentinelUserData
    let deviceData: SentinelDeviceData
    let locationData: SentinelLocationData
    let installedApps: SentinelInstalledApps

    enum CodingKeys: String, CodingKey {
        case metadata
        case userData = "user_data"
        case deviceData = "device_data"
        case locationData = "location_data"
        case installedApps = "installed_apps"
    }
}

File: android/sentinel-core/src/main/kotlin/com/sentinel/sdk/SentinelSDK.kt
package com.sentinel.sdk

import android.content.Context
import android.os.Build
import com.sentinel.sdk.collectors.AppDetectionCollector
import com.sentinel.sdk.collectors.DeviceDataCollector
import com.sentinel.sdk.collectors.LocationCollector
import com.sentinel.sdk.collectors.PersonalDataValidator
import com.sentinel.sdk.models.SentinelDeviceData
import com.sentinel.sdk.models.SentinelInstalledApps
import com.sentinel.sdk.models.SentinelLocationData
import com.sentinel.sdk.models.SentinelMetadata
import com.sentinel.sdk.models.SentinelPayload
import com.sentinel.sdk.models.SentinelUserData
import com.sentinel.sdk.security.PayloadSigner
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID

class SentinelSDK private constructor(private val context: Context) {

    companion object {
        const val VERSION = "1.0.0"

        @Volatile
        private var INSTANCE: SentinelSDK? = null

        val compactJson = Json {
            encodeDefaults = true
            ignoreUnknownKeys = true
            prettyPrint = false
        }

        val prettyJson = Json {
            encodeDefaults = true
            ignoreUnknownKeys = true
            prettyPrint = true
        }

        fun formatJson(payload: SentinelPayload, prettyPrint: Boolean = true): String {
            val serializer = if (prettyPrint) prettyJson else compactJson
            return serializer.encodeToString(payload)
        }

        fun initialize(context: Context, apiKey: String, environment: String = "production"): SentinelSDK {
            return INSTANCE?.apply {
                this.apiKey = apiKey
                this.environment = environment
            } ?: synchronized(this) {
                INSTANCE ?: SentinelSDK(context.applicationContext ?: context).also {
                    it.apiKey = apiKey
                    it.environment = environment
                    it.isInitialized = true
                    INSTANCE = it
                }
            }
        }

        fun getInstance(): SentinelSDK {
            return INSTANCE ?: throw IllegalStateException("SentinelSDK must be initialized first via SentinelSDK.initialize(context, apiKey).")
        }

        fun resetForTesting() {
            synchronized(this) {
                INSTANCE = null
            }
        }
    }

    private var apiKey: String? = null
    private var environment: String = "production"
    private var isInitialized: Boolean = false

    private val personalDataValidator = PersonalDataValidator()
    private val deviceDataCollector = DeviceDataCollector(context)
    private val locationCollector = LocationCollector(context)
    private val appDetectionCollector = AppDetectionCollector(context)

    data class CaptureOptions(
        val userData: PersonalDataValidator.Input,
        val locationTimeoutMs: Long = 5000L,
        val wrapper: String = "native",
        val customTargets: List<AppDetectionCollector.TargetApp>? = null
    )

    suspend fun capture(options: CaptureOptions): SentinelPayload = coroutineScope {
        val startTime = System.currentTimeMillis()

        // 1. Personal Data Validation (CPU-bound)
        val userData = personalDataValidator.validate(options.userData)

        // 2. Parallel Telemetry Collectors
        val deviceDataDeferred = async(Dispatchers.IO) { deviceDataCollector.collect() }
        val locationDataDeferred = async(Dispatchers.IO) {
            locationCollector.collectLocation(timeoutMs = options.locationTimeoutMs)
        }
        val targets = options.customTargets ?: AppDetectionCollector.DEFAULT_TARGETS
        val installedAppsDeferred = async(Dispatchers.IO) {
            appDetectionCollector.scan(targets)
        }

        val deviceData = deviceDataDeferred.await()
        val locationData = locationDataDeferred.await()
        val installedApps = installedAppsDeferred.await()

        val duration = maxOf(1L, System.currentTimeMillis() - startTime)
        val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }

        val metadata = SentinelMetadata(
            sdkVersion = VERSION,
            platform = "android",
            osVersion = Build.VERSION.RELEASE ?: "unknown",
            wrapper = options.wrapper,
            sessionId = UUID.randomUUID().toString(),
            requestId = "req_" + UUID.randomUUID().toString().replace("-", "").lowercase(),
            timestamp = sdf.format(Date(startTime)),
            captureDurationMs = duration
        )

        // Security signature (Canonical compact JSON)
        val unsignedPayload = PreSecurityPayload(
            metadata = metadata,
            userData = userData,
            deviceData = deviceData,
            locationData = locationData,
            installedApps = installedApps
        )

        val isTampered = deviceData.isJailbrokenOrRooted || locationData.isMockLocation
        val signer = PayloadSigner(apiKey)
        val unsignedCanonicalJson = compactJson.encodeToString(unsignedPayload)
        val securityData = signer.sign(unsignedCanonicalJson, tamperDetected = isTampered)

        SentinelPayload(
            metadata = metadata,
            userData = userData,
            deviceData = deviceData,
            locationData = locationData,
            installedApps = installedApps,
            security = securityData
        )
    }

    suspend fun captureJson(options: CaptureOptions, prettyPrint: Boolean = true): String {
        val payload = capture(options)
        val serializer = if (prettyPrint) prettyJson else compactJson
        return serializer.encodeToString(payload)
    }
}

@Serializable
private data class PreSecurityPayload(
    @SerialName("metadata") val metadata: SentinelMetadata,
    @SerialName("user_data") val userData: SentinelUserData,
    @SerialName("device_data") val deviceData: SentinelDeviceData,
    @SerialName("location_data") val locationData: SentinelLocationData,
    @SerialName("installed_apps") val installedApps: SentinelInstalledApps
)

